spec/abci/abci++_app_requirements.md

```diff
@@ -188,6 +188,37 @@ Likewise, `ExtendVote` can also be non-deterministic:
 * *w<sup>r</sup><sub>p</sub> = w<sup>r</sup><sub>q</sub> &#8655;
   e<sup>r</sup><sub>p</sub> = e<sup>r</sup><sub>q</sub>*
 
+Unlike all requirements described above,
+which affect the ABCI methods of the [Consensus Connection](#consensus-connection),
+the following requirement is on the [Mempool Connection](#mempool-connection).
+Let *CheckTxCode<sub>tx,p,t</sub>* denote the error code returned by *p*'s Application, via `ResponseCheckTx`,
+to CometBFT's `RequestCheckTx` call occurring at time *t* and having transaction *tx* as parameter.
+Let predicate *OK(CheckTxCode)* denote whether *CheckTxCode* is `SUCCESS`.
+
+
+* Requirement 13 [`CheckTx`, finite-oscillation]: For any correct process *p*
+  and any transaction *tx*,
+  
+    - (a) either *OK(CheckTxCode<sub>tx,p,t<sub>0</sub></sub>) = OK(CheckTxCode<sub>tx,p,t<sub>1</sub></sub>)*
+      for any two times *t<sub>0</sub>* and *t<sub>1</sub>* in *p*'s execution, or
+    - (b) there exists a time *t<sub>stable</sub>* in *p*'s execution
+      such that *&#172; OK(CheckTxCode<sub>tx,p,t</sub>)* for all *t > t<sub>stable</sub>*.
+
+Requirement 13 has two main conditions. The implementation of `CheckTx` must fulfill at least one of them.
+Condition (a) mandates that `CheckTx` on a transaction never changes its result,
+in other words, that the checks implemented are stateless (e.g., well-formedness, signature verification).
```

Comment:
` My understanding is that it should be disabled for app that do stateless CheckTx.`

If this is indeed the rule, I believe it should be written down. Maybe it is and I just have not read it. Beacuse otherwise you end up with questions such as mine :) 

Then if `RecheckTx` is enabled, I feel that the check is stateless within a height (because you are not supposed to change the state in the first place), but then the requirement a) should be relaxed such that it is allowed to not succeed only if a block was committed.

Reply:
> should the spec consider CheckTx and RecheckTx as different? 

Yes, in my opinion.

> If yes, is there any requirement we want to impose in how they must relate?

I agree with Jasmina's comments below. If the state of the application hasn't change between `CheckTx` and `ReCheckTx` for the same transaction, their output should be the same. In other words, in the absence of state changes (probably we can refer this using heights to indicate state changes), `ReCheckTx` should be handled as it was a duplicated/second call to `CheckTx`.

Reply:
Regarding the text specifically, I think that condition (a) prevents the application to implement replay protection mechanisms. And our assumption is kind of that is _mandatory_ the application to implement such mechanisms.

I wonder how we can state condition (a) in a way it still allows the application to reject `tx` in the case the same `tx` has already been included in a committed block.